File: library/builtins/traits/number.rs
pub use rust_decimal;

use crate::*;
use rust_decimal::Decimal;
use wipple::*;

#[derive(TypeInfo, Debug, Clone)]
pub struct Number {
    pub number: Decimal,
    pub location: Option<SourceLocation>,
}

impl Primitive for Number {}

impl Number {
    pub fn new(number: Decimal) -> Self {
        Number::new_located(number, None)
    }

    pub fn new_located(number: Decimal, location: Option<SourceLocation>) -> Self {
        Number { number, location }
    }
}

pub(crate) fn setup(env: &Env, stack: &Stack) -> Result<()> {
    env.set_variable("Number", Value::of(Trait::of::<Number>()));

    // Number == Text
    env.add_relation_between(stack, |number: Number| {
        Text::new(number.number.normalize().to_string())
    })?;

    Ok(())
}

File: Library/Sources/Traits/Number.swift
import Foundation
import Wipple

public struct Number: RawRepresentable, Primitive {
    public var rawValue: Decimal

    public init(rawValue: Decimal) {
        self.rawValue = rawValue
    }
}

internal func setupNumber(_ env: Env, _ stack: Stack) throws {
    try env.setVariable("Number", to: Value(Trait(Number.self)), stack)

    // Number == Text
    try env.addRelation(from: Number.self, to: Text.self, stack) { number in
        Text(rawValue: "\(number.rawValue)")
    }
}
